### DISCUSSION TITLE
Transaction boundaries

### DISCUSSION THREAD

**Discussion by matthooper** [lecaillon/Evolve#209] [Q&A]
What happens if during a single CLI migration run, several scripts succeed and then a subsequent one fails? Are all changes from the successful scripts in that migration run rolled back or only the changes in the failed script rolled back?

Thanks!

**Comment by lecaillon** [comment]
Each migration is executed in a separate database transaction. Thus each script will either succeed or fail completely and Evolve will stop on the first error ([here](https://evolve-db.netlify.app/concepts/#transactions) in the documentation).

So all changes from the successful scripts are applied.

**Comment by matthooper** [comment]
Thanks for the fast reply!

Not having a parent transaction would be a roadblock for us adopting
Evolve. How feasible would adding a parent transaction be? Either
optionally or always?

**Reply by lecaillon** [reply]
Indeed it could be a good feature to add as an option. Let me check what it costs to add it in the code base. In the meantime do you have a suggestion for the name of the option ? (here is the list of existing ones : https://evolve-db.netlify.app/configuration/#options)

And to help me prioritize tests, what database would you use with Evolve ?

If it appears to be feasible (either clean, testable, xplat) I will do my best to develop it as soon as possible. At least in a alpha [release](https://github.com/lecaillon/Evolve/releases) (3.0.0-alpha5 for example)

What do you think ?

**Comment by matthooper** [comment]
That sounds great!

The database would be SQL Server (>=2014, Standard Edition).

Some name ideas:

   - Transaction, TransactionLevel, TransactionScope = {Command, Migration}
   or {All, Each} or {CommitAll, CommitEach}

Also, if there could be an option or command to trigger a rollback without
error after the last successful script, that would be very useful. That way
upgrades could be tested against an existing database prior to doing the
real thing. Also useful to verify new SQL scripts automatically in a CI
setup or before running in a high-risk production environment. That could
either be a new command (ideas: test-migrate) or an option (ideas:
Test=true, Transaction=RollbackAll).

My vote would be Transaction={CommitAll, CommitEach, RollbackAll}.

Happy to talk more if needed. Thanks for your help!

**Reply by lecaillon** [reply]
Great ideas, great naming, I'll keep you in touch

**Reply by lecaillon** [reply]
Option name : **Evolve.TransactionMode**
- CommitEach (default)
- CommitAll or CommitOnce (?)
- RollbackAll

**Comment by lecaillon** [comment]
**Evolve 3.0.0-alpha5** is released. Feel free to test the new option `TransactionMode` either with value `CommitAll ` or `RollbackAll`.
Waiting for your feedback @matthooper

**Comment by matthooper** [comment]
Thanks! I skimmed the changes and didn't see anything that looked off. I'm
working to get into a sprint for my team to review as part of their tool
evaluation.

**Comment by matthooper** [comment]
The new TransactionMode options seems to be working great from the CLI.
Thanks again!
